### ISSUE TITLE
Element div not allowed as child of element button in this context

### ISSUE THREAD

**Issue by hrvojegolcic** [validator/validator#1332]
Hi there, the following code is being validated:

```html
<!DOCTYPE html>
<html lang="">
<head>
<title>Test</title>
</head>
<body>

<button><div>INVALID</div></button>

<a role="button"><div>SHOULD BE INVALID THEN TOO</div></a>

</body>
</html>
```

As per error message, both cases should be invalid then, but only the first one currently is.

**Comment by scottaohara** [comment]
not exactly.

a `div` is not allowed a child/descendant of a `button` _element_

a `div` is an allowed child/descendant of an `a` _element_.

an `a role=button` is still an `a` element in how it is rendered by the browser, but its implicit aria role has been overwritten by the `button` role.

**Comment by hrvojegolcic** [comment]
Well, at least JAWS and VoiceOver will treat `role=button` same as `<button>` in my case if I put a heading instead of DIV, it won't be recognized as a heading if there is `role=button` or ˙<button>˙.

Nonetheless then there is some inconsistency here since, in similar case there is specific complain about the lement with the role itself.

```html
<!DOCTYPE html>
<html lang="">
<head>
<title>Test</title>
</head>
<body>

<a><div tabindex="-1">VALID</div></a>
<a role=button><div tabindex="-1">VALID</div></a>

</body>
</html>
```

![image](https://user-images.githubusercontent.com/6906827/157702170-6bd94aa1-ab40-4ef9-9f98-1fd59c7a82a1.png)

**Comment by scottaohara** [comment]
>Well, at least JAWS and VoiceOver will treat role=button same as <button>, which is the whole point of that role after all.

they announce the role the same.  They don't do anything beyond that.  E.g., a `div role=button` requires developers to provide all the necessary functionality.  But this validation flag is not about exposed roles, but rather HTML's allowed nesting of elements.   

re: nested headings - yeh... that's something that need to be further defined and is actually an open issue for ARIA in HTML to specify. While related, that's beyond the scope of what's being called out here.

re: your final example about element with tabindex=-1, again that's a different check about not having nested focusables in interactive elements.

Again, i would suggest that this is more a topic of "is it really a problem for a div to be a child of an html button?"  in the past it likely was, but browsers do pretty well at error correcting for that nesting these days.

i don't see the point in flagging the following though:
```html
<div role=button tabindex=0>
  <div style=...>...</div>
</div>
```

so long as there are no other roles/properties important for accessibility, there's nothing 'wrong' with the markup snippet i provided.  saying that a `div` cannot be allowed as a descendant of `role=button` is conflating HTML element parsing requirements with accessibility concerns, which are definitely related at times, but don't always cause 1 to 1 issues.

**Comment by hrvojegolcic** [comment]
Surely I agree with you in regards of `is it really a problem for a div to be a child of an html button?` clarification. I wasn't concerned only about `div`, but generally headings too. My point goes to the consistency. Imo wherever `role="button"` is used one should be able to use its semantic equivalent `<button>` instead and same rules should apply. Am I getting this wrong?
  
If so, maybe then instead, the issue is rather that a new validation rule should be introduced, such as "An element with `role=button`" must not contain a heading. Because [accessibility APIs do not have a way of representing semantic elements contained in a button](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#all_descendants_are_presentational) (hence probably a reason why JAWS and VoiceOver will ignore the heading in this case). That would be a different issue then the one reported here so if you agree we can close this one and open another one. Unless we should indeed have same validation rules for `<button>` and `role=button`

The second case is solved for `button` but not for `role=button` as follows:

```html
<!DOCTYPE html>
<html lang="">
<head>
<title>Test</title>
</head>
<body>

<button><div role="heading" aria-level="2">I am a heading</div></button>

<button><h2>I am a heading</h2></button>

<div role="button"><div role="heading" aria-level="2">I am a heading</div></div>

<div role="button"><h2>I am a heading</h2></div>

</body>
</html>
```

![image](https://user-images.githubusercontent.com/6906827/158032864-bf4e4a52-098c-4529-b14c-a8d2afea95da.png)

**Comment by scottaohara** [comment]
>Imo wherever role="button" is used one should be able to use its semantic equivalent <button> instead and same rules should apply. Am I getting this wrong?

Yes.

Again, the no `div` or `h#` element inside of a `button` element rule is due to HTML's parsing requirements.  The only elements allowed within a `<button>` are [phrasing content elements](https://html.spec.whatwg.org/multipage/dom.html#phrasing-content-2) but there must be no [interactive content](https://html.spec.whatwg.org/multipage/dom.html#interactive-content-2) descendant and no descendant with the [tabindex](https://html.spec.whatwg.org/multipage/interaction.html#attr-tabindex) attribute specified.

The reason a heading's implicit ARIA role is being suppressed within a `<button>` or `role=button` is because a button role treats child elements as if they were `role=presentation`.  There is no requirement, nor should there be, about whether a `div` is an allowed descendant of `role=button`.  

The rules you are looking for would be normatively defined by the resolution of https://github.com/w3c/html-aria/issues/362

**Comment by hrvojegolcic** [comment]
> The reason a heading's implicit ARIA role is being suppressed

Yes, that's exactly what the MDN article I posted says

> The rules you are looking for would be normatively defined by the resolution of [w3c/html-aria#362](https://github.com/w3c/html-aria/issues/362)

Thank you. Not sure how to understand this. Is it that per your opinion `role` thingies should not be part of HTML validation at all?

**Comment by scottaohara** [comment]
no. that's not what i've been saying.  If you read the ARIA in HTML spec, you'd see that there are many rules for use of ARIA with HTML that **are** presently part of validating HTML.

What I have been telling you is that nesting allowances for HTML elements and nesting allowances for HTML elements with explicit ARIA roles are not a 1 to 1 match.  Your original post in this thread is one example of this discrepancy, because there is nothing wrong with it from an HTML validation standpoint, nor a validation of HTML & ARIA.

An `<a>` element, with or without an `href` allows for a `<div>` as a child.  So no failure from HTML.

An `<a role=button>`, with or without an `href` also allows for a `<div>` as a child because the parsing rules for the `<a>` element still apply, and from an ARIA standpoint, `<div>` is a `generic` element, exposing no meaningful semantics.  So, because a `role=button` treats children as pesentational (or mostly does - that's a whole other rabbit hole) there is nothing to flag someone for concerning the `<div>`. 

Your shift to talking about headings within buttons being invalid is something that does need to be addressed with the validator. The ARIA in HTML issue I linked to is where we will be defining the nesting rules to account for situations like:

```html
<button> <div role=heading>invalid</div> </button>
```
where the `div` is invalid due to HTML's nesting rules, but the use of `role=heading` on an element inside of a `<button>` is the rule that ARIA in HTML needs to define.

**Comment by hrvojegolcic** [comment]
> something that does need to be addressed with the validator

I see clearly now where we have a disagreement on the matter. You don't, but I do however consider that it would be useful for the validator to address for `role=button`, the same way as it does for `<button>`. If nothing, at least for consistency. That is my opinion and I stay with it. Surely if the maintainers don't find this useful can close the issue. Thank you for your inputs.

**Comment by scottaohara** [comment]
What value is consistency if the validation error isn’t actually an error?

**Comment by hrvojegolcic** [comment]
Everything explained in the first post. I agree we disagree and this
discussion doesn't lead anywhere anymore. So thank you for pointing out
what's on your mind. And likewise.

**Comment by sideshowbarker** [comment]
@scottaohara Is there anything here for which the checker is out of conformance with current specs?

**Comment by sideshowbarker** [comment]
Closing this since as far as I can glean from the discussion, the checker behavior on this isn’t out of conformance with any of the relevant specs. (But @scottaohara and @pkra if I’m wrong, lemme know.)
